File: day03/aoc3-1.pl
#!/usr/bin/perl
use strict;
use warnings;

sub largest_bank_number {
    my ($line) = @_;
    my @digits = grep { /^\d$/ } split //, $line;

    return -1 if @digits < 2;

    # Build suffix max array: max digit strictly to the right of position i
    my @sufmax = (-1) x scalar(@digits);
    my $current_max = -1;
    for (my $i = $#digits; $i >= 0; $i--) {
        $sufmax[$i] = $current_max;
        $current_max = $digits[$i] if $digits[$i] > $current_max;
    }

    my $best_val = -1;
    my $best_str = "-1";

    for my $i (0 .. $#digits - 1) {
        my $right_max = $sufmax[$i];
        next if $right_max == -1;  # nothing to the right
        my $val = $digits[$i] * 10 + $right_max;
        if ($val > $best_val) {
            $best_val = $val;
            $best_str = $digits[$i] . $right_max;
        }
    }

    return $best_val;  # return the two-digit number as an integer
}

# Read from aocdata.txt
my $filename = "aocdata.txt";
open my $fh, '<', $filename or die "Could not open '$filename': $!";

my $sum = 0;

while (my $line = <$fh>) {
    chomp $line;
    next if $line =~ /^\s*$/;  # skip blank lines
    my $largest = largest_bank_number($line);
    print "Bank: $line -> Largest possible number: $largest\n" if $largest != -1;
    $sum += $largest if $largest != -1;
}

close $fh;

print "Total sum of largest numbers: $sum\n";

File: old/day03/aoc3-1.py
def largest_bank_number_two(line: str) -> int:
    digits = [int(ch) for ch in line if ch.isdigit()]
    n = len(digits)
    if n < 2:
        return -1

    # Suffix max: for each i, the largest digit strictly to its right
    sufmax = [-1] * n
    cur = -1
    for i in range(n - 1, -1, -1):
        sufmax[i] = cur
        if digits[i] > cur:
            cur = digits[i]

    best_val = -1
    for i in range(n - 1):  # i can be up to n-2
        right_max = sufmax[i]
        if right_max == -1:
            continue
        val = digits[i] * 10 + right_max
        if val > best_val:
            best_val = val

    return best_val


def process_file_two(filename: str):
    total = 0
    with open(filename) as f:
        for line in f:
            s = line.strip()
            if not s:
                continue
            val = largest_bank_number_two(s)
            print(f"Bank: {s} -> Largest possible 2-digit number: {val}")
            if val != -1:
                total += val
    print(f"Total sum of largest 2-digit numbers: {total}")


# Example usage
process_file_two("aocdata.txt")
